tests/Test/Xrefcheck/TooManyRequestsSpec.hs

```diff
@@ -0,0 +1,59 @@
+{- SPDX-FileCopyrightText: 2021 Serokell <https://serokell.io>
+ -
+ - SPDX-License-Identifier: MPL-2.0
+ -}
+
+module Test.Xrefcheck.TooManyRequestsSpec where
+
+import Universum
+
+import Control.Concurrent (forkIO, killThread)
+import qualified Control.Exception as E
+import qualified Data.CaseInsensitive as CI
+import qualified Data.Map as M
+import Network.HTTP.Types (Status (..), ok200, serviceUnavailable503, tooManyRequests429)
+import Network.HTTP.Types.Header (hRetryAfter)
+import Test.Hspec (Spec, describe, it, shouldBe)
+import Web.Firefly (ToResponse (toResponse), route, run)
+
+import Xrefcheck.Config
+import Xrefcheck.Core
+import Xrefcheck.Verify
+
+spec :: Spec
+spec = do
+  describe "429 response tests" $ do
+    it "Returns 200 eventually" $
+      checkLinkWithServer (mock429 True)
+        "http://127.0.0.1:5000/429" $ VerifyResult []
+    it "Returns 503 eventually" $
+      checkLinkWithServer (mock429 False)
+        "http://127.0.0.1:5000/429" $ VerifyResult
+          [ ExternalHttpResourceUnavailable $
+              Status { statusCode = 503, statusMessage = "Service Unavailable"}
+          ]
+  where
+    checkLinkWithServer mockServer link expectation =
+      E.bracket (forkIO (newIORef 0 >>= mockServer)) killThread $ \_ -> do
+        verifyReferenceDef link >>= (`shouldBe` expectation)
+
+    verifyReferenceDef link = do
+      progRef <- newIORef $ initVerifyProgress []
+      fmap wrlItem <$> verifyReference
+        ((cVerification $ defConfig GitHub) { vcCheckLocalhost = True }) FullMode
+        progRef (RepoInfo M.empty) link ""
+        (Reference "" (toText link) Nothing (Position Nothing))
+
+    mock429 :: Bool -> IORef Int -> IO ()
+    mock429 mode switchRef = run 5000 $ do
+      route "/429" $ do
```

Comment:
The IORef is always initialized to 0, so maybe we can call the `newIORef` when starting up the server instead of passing it as a parameter.

```suggestion
    mock429 :: Bool -> IO ()
    mock429 mode = do
      switchRef <- newIORef 0
      run 5000 $ do
        route "/429" $ do
```

Reply:
I have tried bringing `switchRef` from the global list of arguments to the definition of `mock429`, but that seemed to have been breaking the tests: the actual `VerifyResult` values contain `ExternalHttpTooManyRequests`, meaning we've exceeded the maximum number of retries both times.
![image](https://user-images.githubusercontent.com/40356589/141860534-0a706dbc-9597-477e-ad8a-0b11f13dab9d.png)

Reply:
Did you call `newIORef` before or after `run 5000`?

Reply:
Before.

Reply:
Hm weird, I tried this and all the tests still pass..

![image](https://user-images.githubusercontent.com/718166/146739577-f17a4844-f054-44ae-9085-10918b7f2fbf.png)

I don't see a reason why it wouldn't... Would you mind giving this another try please?

Reply:
Tried it again, and this time it worked! Turns out I have put the `IORef` initialization _after_ `run 5000` and _before_ 
`route "/429"`; and I believe writing `run 5000` on the same line as the definition of `mock429` could have possibly led to the confusion on my part.